docs/utxo-hd/v0.1/for-developers.md

```diff
@@ -0,0 +1,276 @@
+
+- [InMemory](../../../ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/LedgerDB/BackingStore/InMemory.hs):
+  is a `Data.Map` that lives in a `TVar`.
+- [LMDB](../../../ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/LedgerDB/BackingStore/LMDB.hs)
+  uses [LMDB](http://www.lmdb.tech/doc/) bindings to keep a database in a file in the disk.
+
+To be precise, only the LMDB implementation will decrease the memory usage, but
+it comes at a cost as I/O has to be performed to read and write values. We
+expect SPOs with powerful machines to use the InMemory implementation and every
+other user to move to the LMDB implementation.
+
+### Communication with the ledger
+
+From the point of view of the communication with the Ledger rules, it turns out
+that for each block we need a way to inject (and project) the ledger tables into
+the actual ledger state so that we can continue using the Ledger code without
+modifying it. As the Ledger ledger-state type depends on the type of the block,
+each block type defined in the Consensus layer has to provide a way to do so.
+
+But the Consensus-Ledger API assumes the existence of this projection/injection
+functions for each block type (i.e. needs the instance `CanStowLedgerTables
+(LedgerState blk)`), and it defines a general flow for **applying a block: a
+ledger state with some values (`ValuesMK`) goes in and a ledger state with the
+corresponding deltas (`DiffMK`) comes out**. This is reflected in the
+`(re)applyBlockLedgerResult` functions:
+
+```haskell
+applyBlockLedgerResult ::
+     HasCallStack
+  => LedgerCfg l
+  -> blk
+  -> Ticked1 l ValuesMK
+  -> Except (LedgerErr l) (LedgerResult l (l DiffMK))
+```
+
+### Preparing a ledger state for the Ledger rules
+
+In order to provide a `LedgerState` with values (`ValuesMK`) to the Ledger, we
+need to follow this sequence of steps:
+
+1. Consult which UTxOs the block will need (`getBlockKeySets`).
+
+2. Read those from the BackingStore (module `ReadsKeySets`).
+
+3. Apply the diffs in the sequence of deltas in memory up to the desired slot (
+   `forwardTableKeySets`).
+
+4. Inject the deltas into the ledger state and call the desired Ledger rule.
+
+This read and forwarding is another of the main sources of speed loss in the
+node. Before UTxO-HD, the values were just in memory already so we didn't need
+to do any further work to bring them in. Now we even have (depending on the
+backing store chosen) to perform I/O to read those.
+
+### Era boundaries
+
+In general, ticking a ledger state should not modify the values in the UTxO set,
+but there are two exceptions in the Cardano case when ticking crosses an era
+boundary. For this reason, the general scheme for **ticking a block is: a ledger
+state without a UTxO set (`EmptyMK`) goes in and a ledger state with some deltas
+(`DiffMK`) comes out**. This is reflected in `applyChainTickLedgerResult`.
```

Comment:
Maybe it's worth repeating that the deltas of `applyChainTickLedgerResult` will be empty except in two cases/era boundaries (mentioned above).